src/core/thread/mesh_forwarder.cpp

```diff
@@ -2085,6 +2094,20 @@ void MeshForwarder::HandleFragment(uint8_t *aFrame, uint8_t aFrameLength,
             }
         }
 
+        // For a sleepy-end-device, if we receive a new (secure) next fragment
+        // with a non-matching fragmentation offset or tag, it indicates that
+        // we have either missed a fragment, or the parent has moved to a new
+        // message with a new tag. In either case, we can safely clear any
+        // remaining fragments stored in the reassembly list.
+
+        if (GetRxOnWhenIdle() == false)
+        {
+            if ((message == NULL) && (aMessageInfo.mLinkSecurity))
```

Comment:
This PR looks good.  It isn't really clear how this line performs the check outlined in the comment however. `(message == NULL)` somehow signifies a missed fragment or a new tag?